// react-hardware/lib/components/RGBLed.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Led = require('./Led');

var _Led2 = _interopRequireDefault(_Led);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RGBLed = function (_Component) {
  _inherits(RGBLed, _Component);

  function RGBLed() {
    _classCallCheck(this, RGBLed);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(RGBLed).apply(this, arguments));
  }

  _createClass(RGBLed, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var pins = _props.pins;
      var values = _props.values;


      return _react2.default.createElement(
        'container',
        null,
        pins.map(function (pin, i) {
          return _react2.default.createElement(_Led2.default, {
            pin: pin,
            value: typeof values[i] !== 'undefined' ? values[i] : values[0],
            mode: 'PWM'
          });
        })
      );
    }
  }]);

  return RGBLed;
}(_react.Component);

RGBLed.propTypes = {
  pins: _react.PropTypes.arrayOf(_react.PropTypes.number).isRequired,
  values: _react.PropTypes.arrayOf(_react.PropTypes.number).isRequired
};

exports.default = RGBLed;
